File: examples/ledger/python/ledger_cli.py
from __future__ import annotations

import argparse
import os
import sys

from bowline_client import BowlineError, SyncTransport

from bowline import (
    CreateInvoiceInput,
    GetInvoiceInput,
    Invoice,
    InvoiceLocked,
    Line,
    ListInvoicesInput,
    SyncClient,
    VoidInvoiceInput,
)


def connect() -> SyncClient:
    url = os.environ.get("LEDGER_URL", "http://localhost:8080/api")
    token = os.environ.get("LEDGER_TOKEN")
    headers = (lambda: {"authorization": f"Bearer {token}"}) if token else None
    return SyncClient(SyncTransport(url, headers=headers))


def show(invoice: Invoice) -> None:
    created = invoice.createdAt.astimezone().strftime("%Y-%m-%d %H:%M")
    print(f"{invoice.id:>4}  {invoice.status.value:<6} {invoice.total:>14}  {created}")


def run(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(prog="ledger", description="Talk to the ledger example.")
    commands = parser.add_subparsers(dest="command", required=True)
    commands.add_parser("list", help="list invoices")
    create = commands.add_parser("create", help="create an invoice for customer 1")
    create.add_argument("description")
    create.add_argument("quantity", type=int)
    void = commands.add_parser("void", help="void an invoice")
    void.add_argument("id", type=int)
    get = commands.add_parser("get", help="show one invoice")
    get.add_argument("id", type=int)
    args = parser.parse_args(argv)
    client = connect()
    try:
        if args.command == "list":
            page = client.invoices.list(ListInvoicesInput(limit=50))
            for invoice in page.items:
                show(invoice)
        elif args.command == "create":
            line = Line(description=args.description, quantity=args.quantity, unitPrice="USD 10.00")
            show(client.invoices.create(CreateInvoiceInput(customerId=1, lines=[line])))
        elif args.command == "void":
            show(client.invoices.void(VoidInvoiceInput(id=args.id)))
        elif args.command == "get":
            show(client.invoices.get(GetInvoiceInput(id=args.id)))
    except BowlineError as err:
        if err.type == "InvoiceLocked":
            locked = err.details_as(InvoiceLocked)
            print(
                f"invoice {locked.id} is locked because it is {locked.status.value}",
                file=sys.stderr,
            )
        else:
            print(f"{err.code.value}: {err.message}", file=sys.stderr)
            for issue in err.issues:
                print(f"  {'.'.join(issue.path)}: {issue.message}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(run(sys.argv[1:]))

File: examples/ledger/elixir/lib/ledger_cli.ex
defmodule LedgerCli do
  @moduledoc "Lists, creates, and voids ledger invoices through the generated Elixir client."

  alias BowlineClient.Error
  alias Ledger.Invoices
  alias Ledger.Types

  @spec transport() :: BowlineClient.Transport.t()
  def transport do
    Ledger.Client.new(System.get_env("LEDGER_URL", "http://localhost:8080/api"))
  end

  @spec list(BowlineClient.Transport.t()) :: :ok
  def list(transport) do
    case Invoices.list(transport, %Types.ListInvoicesInput{limit: 20}) do
      {:ok, page} -> Enum.each(page.items, &print/1)
      {:error, error} -> fail(error)
    end
  end

  @spec create(BowlineClient.Transport.t(), String.t(), integer()) :: :ok
  def create(transport, description, quantity) do
    input = %Types.CreateInvoiceInput{
      customer_id: 1,
      lines: [%Types.Line{description: description, quantity: quantity, unit_price: "USD 10.00"}]
    }

    case Invoices.create(transport, input) do
      {:ok, invoice} -> print(invoice)
      {:error, error} -> fail(error)
    end
  end

  @spec void(BowlineClient.Transport.t(), integer()) :: :ok
  def void(transport, id) do
    case Invoices.void(transport, %Types.VoidInvoiceInput{id: id}) do
      {:ok, invoice} -> print(invoice)
      {:error, error} -> fail(error)
    end
  end

  defp print(%Types.Invoice{} = invoice) do
    created = Calendar.strftime(local_time(invoice.created_at), "%Y-%m-%d %H:%M")
    IO.puts("#{invoice.id}\t#{invoice.status}\t#{invoice.total}\t#{created}")
  end

  defp fail(%Error{} = error) do
    IO.puts(:stderr, Exception.message(error))
    exit({:shutdown, 1})
  end

  defp local_time(%DateTime{} = utc) do
    utc
    |> DateTime.to_naive()
    |> NaiveDateTime.to_erl()
    |> :calendar.universal_time_to_local_time()
    |> NaiveDateTime.from_erl!()
  end
end

defmodule Mix.Tasks.Ledger.List do
  @moduledoc "Lists invoices: mix ledger.list"
  @shortdoc "List invoices"
  use Mix.Task

  @impl true
  def run(_args) do
    Mix.Task.run("app.start")
    LedgerCli.list(LedgerCli.transport())
  end
end

defmodule Mix.Tasks.Ledger.Create do
  @moduledoc "Creates an invoice: mix ledger.create <description> <quantity>"
  @shortdoc "Create an invoice"
  use Mix.Task

  @impl true
  def run([description, quantity]) do
    Mix.Task.run("app.start")
    LedgerCli.create(LedgerCli.transport(), description, String.to_integer(quantity))
  end

  def run(_), do: Mix.raise("usage: mix ledger.create <description> <quantity>")
end

defmodule Mix.Tasks.Ledger.Void do
  @moduledoc "Voids an invoice: mix ledger.void <id>"
  @shortdoc "Void an invoice"
  use Mix.Task

  @impl true
  def run([id]) do
    Mix.Task.run("app.start")
    LedgerCli.void(LedgerCli.transport(), String.to_integer(id))
  end

  def run(_), do: Mix.raise("usage: mix ledger.void <id>")
end
